metadata/src/test/java/org/apache/kafka/controller/MockMetaLogManagerListener.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.controller;
+
+import org.apache.kafka.common.protocol.ApiMessage;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+public class MockMetaLogManagerListener implements MetaLogManager.Listener {
+    private final AtomicReference<String> firstError;
+    private final CommitHandler commitHandler;
+    private final int nodeId;
+    private boolean shutdown = false;
+    private long curEpoch = -1;
+
+    interface CommitHandler {
+        void handle(int nodeId, long epoch, long index, ApiMessage message) throws Exception;
+    }
+
+    MockMetaLogManagerListener(AtomicReference<String> firstError,
```

Comment:
> public rather than package-private, or is it this way on purpose?

It is this way on purpose, since this is only used in tests for this package

> Also, should firstError be taken from a value passed in

The intention is to have only one of these for the whole test, to avoid having to do lots of checks.  So it's a shared atomic reference.

> Actually, since the compareAndSet Javadoc says Atomically sets the value to the given updated value if the current value == the expected value I'm thinking null might be better than ""

Good catch!  I guess I need to read the docs more closely.  null seems like a reasonable choice here.  A sentinel value would be fine too, but it's a bit too clever, and sometimes using == with strings trips code linters.

Reply:
I changed this to use null instead of the empty string.